### PR TITLE
My HW-03 tasks

### PR DESCRIPTION
**Pull request by ExileeY**

### CODE REVIEW

File: HW03/Danila_Glazov/danila_glazov_hw_03_t_02.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+def format_logs(logs_str)
+  valid_format = /^\d{2}\.\d{1}\.\d{3}\.\d{3}\s-\s-\s.+$/
+  formated_logs = []
+
+  logs_str.split("\n").each do |log|
+    next unless log.match?(valid_format)
+
+    data = log.match(%r{\d{1,2}/\w{3}/\d{4}:\d{2}:\d{2}:\d{2}\s\+\d{4}})
+    from = log.match(/^\d{1,2}\.\d{1}\.\d{1,3}\.\d{1,3}/)
+    to = log.match(%r{"\w+\s(.*)\s\w+/\d{1}\.\d{1}})[1].upcase
```
Comment (B0RlS) [inline HW03/Danila_Glazov/danila_glazov_hw_03_t_02.rb:12]: Could you move regex to constants please?

File: HW03/Danila_Glazov/danila_glazov_hw_03_t_04.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: false
+
+def letter_counter(str)
```
Comment (B0RlS) [inline HW03/Danila_Glazov/danila_glazov_hw_03_t_04.rb:3]: I appreciated your creativity and the fact that you improved the task, but could you also make the task that is specified in your requirements?

File: HW03/Danila_Glazov/danila_glazov_hw_03_t_01.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: false
+
+def find_error(logs)
+  errors = logs.split("\n").select { |log| log.downcase.include?('error') }
+  puts errors.empty? ? '' : errors[0]
```
Comment (vladhilko) [inline HW03/Danila_Glazov/danila_glazov_hw_03_t_01.rb:5]: Maybe?
```suggestion
  logs.split("\n").find { |log| log.downcase.include?('error') }.to_s
```

File: HW03/Danila_Glazov/danila_glazov_hw_03_t_02.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+  logs_str.split("\n").each do |log|
+    next unless log.match?(VALID_FORMAT)
+
+    data = log.match(DATA)
+    from = log.match(FROM)
+    to = log.match(TO)[1].upcase
+
+    formated_logs << "#{data} FROM: #{from} TO: #{to}"
+  end
```
Comment (vladhilko) [inline HW03/Danila_Glazov/danila_glazov_hw_03_t_02.rb:19]: We usually don't use temp variable in ruby, you can try to do something like this 
```suggestion
  logs_str.split("\n").select { |log|  log.match?(VALID_FORMAT) }
    .map do |log|
      data = log.match(DATA)
      from = log.match(FROM)
      to = log.match(TO)[1].upcase

      "#{data} FROM: #{from} TO: #{to}"
  end
```

File: HW03/Danila_Glazov/danila_glazov_hw_03_t_03.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: false
+
+DATETIME = /^\d{4}-\d{1,2}-\d{1,2}\s\d{1,2}:\d{1,2}:\d{1,2}\.\d{1}/.freeze
+
+require 'time'
+def calc_duration(actions)
+  duration_results = []
+  prev_datetime = ''
+
+  actions.split("\n").each do |action|
```
Comment (vladhilko) [inline HW03/Danila_Glazov/danila_glazov_hw_03_t_03.rb:10]: Where do you select only lines with `Calling core with action`? 🤔 
Screenshot 2021-04-28 at 18 16 19

File: HW03/Danila_Glazov/danila_glazov_hw_03_t_04.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: false
+
+DIGITS = /[^\d]/.freeze
+LETTERS = /[^a-zA-Z]/.freeze
+
+def string_info(str)
+  digits = str.gsub(DIGITS, '')
+  letters = str.gsub(LETTERS, '')
```
Comment (vladhilko) [inline HW03/Danila_Glazov/danila_glazov_hw_03_t_04.rb:8]: Interesting approach 👍 But could you use a more clear naming? For example
```suggestion
NOT_DIGITS = /[^\d]/.freeze
NOT_LETTERS = /[^a-zA-Z]/.freeze

def string_info(str)
  digits = str.gsub(NOT_DIGITS, '')
  letters = str.gsub(NOT_LETTERS, '')
```

**Comment by vladhilko** [comment]
Looks good to me, fix 3rd task and I would be happy to approve

**Review by vladhilko — APPROVED** [review]

**Pull request merged**